## Instructions

Answer the following questions. It behooves you to answer as much of each question as possible to receive some partial credit. You are advised to begin this lab early so that you have ample time to complete it.

## Assignment

This assignment is worth 100 points. There are three sections to this homework assignment: a Fundamentals section, an Essentials section, and an Advanced section. You have all of the tools you need to complete all of the lab, but it is triaged for you to give you an idea of what you should focus on first.

### Fundamentals (30 points)

1. (10 points) Explain the difference between an `else` statement and an `elif` statement. When would you use an `else` statement over an `elif` statement?

  > Answer:

2. (10 points) Explain the difference between a `for` loop and a `while` loop. When would you use a `for` loop? A `while` loop?

  > Answer: 

3. (10 points) What do the `break` and `continue` statements do in a loop? Give an example of when you would use each.

 >Answer:

### Essentials (50 points)

4. (10 points) Write an algorithm that prompts the user for 2 text strings.  Then calculate and print the number of matching characters as you progress from the left to the right (from position 0).  For example, if the user provides "hello" and "help" your final answer should be 3.  If the user provides   "hell" and "hello" your final answer should be 4.

    Your algorithm needs to work when comparing same size strings (same length) and differing size strings in any order. 

    >Hint 1:  Use a loop to compare the characters up to length of the shortest string.   Be careful to terminate after exhausting your shorter string.   You MUST NOT continue to compare beyond the last character of the shorter string.
    >Hint 2: You can access individual characters in a string the same way you can access individual items in a list by using `nameOfString[i]`.

In [5]:
## Answer:

str1 = input('Give a string: ')
str2 = input('Give another string: ')

if(len(str1)>len(str2)):
    length = len(str2)
else:
    length = len(str1)

matches = 0
for i in range(length):
    if str1[i] is str2[i]:
        matches+=1

print(matches)

Give a string:  hello
Give another string:  hell


4


5. (10 points) Implement the following algorithm in python.  The algorithm derives the $log_2{(num)}$ by counting how many times we can divide num by 2 until our num shrinks to 1.  Pseudo code is given below.

    + prompt the user for an integer (num)
    + declare an integer (count) and initialize it to be 0
    + while our number (num) is larger than 1 do the following:
         + add 1 to our count variable
         + update our num to be half as big (dividing num by 2 ... integer division)
    + output/print the final value of count.

> For example,  if we start with num = 8.... we will increment our count to 1 and update num to 4 and repeat the body of our while loop.   count increments to 2 and num updates to 2.   Then again, count increments to 3 and num updates to 1.   Since num is no longer > 1 (it is now equal to 1) the while loop exits.  And our algorithm shows the final answer as 3.  In other words, $ 2^3 = 8 $   or  $ log_2{(8)} = 3 $.


In [9]:
## Answer

num = int(input('Give number: '))
count = 0

while(num > 1):
    count += 1
    num /= 2
    
print(count)

Give number:  512


9


6. (10 points) Use iteration to print all of the divisors of a number -- N -- obtained from the human user.   

> Note: X is a **divisor** of N if the remainder is zero after you divide N by X.

In [14]:
## Answer

n = int(input('Give number: '))
x = 1

while(x<=n):
    if(n%x==0):
        print(x)
    x+=1

Give number:  127


1
127


7. (10 points) As activity director at the Austin College Club in Texas, it is your job to suggest appropriate activities to guests based on the weather: 

-  when temp is 80 and above: go swimming in Lake Texoma
-  when temp is 60 and above but less than 80:  play tennis
-  when temp is 40 and above but less than 60:  play golf
-  when temp is less than 40: go skiing on the Grand Ave Overpass

   
    1.	Write a Python algorithm that prompts the user for a temperature, then prints out the activity appropriate for that temperature. Use a cascading **if-elif-else** control structure, and be sure that your conditions are no more complex than necessary.

    2.	Modify your program so that if the temperature if greater than 95 or less than 20, it prints "Visit our campus bookstore!". ( Hint: Use a compound boolean expression )

In [20]:
## Answer Part 1

temp = int(input('What is the temperature? '))

if(temp>=80):
    print('Go swimming in Lake Texoma')
elif(temp<80 and temp>=60):
    print('play tennis')
elif(temp<60 and temp>=40):
    print('play golf')
else:
    print('go skiing on the Grand Ave Overpass')

What is the temperature?  20


go skiing on the Grand Ave Overpass


In [31]:
## Answer Part 2

temp = int(input('What is the temperature? '))

if(temp>95 or temp<20):
    print('visit our campus bookstore!')
elif(temp>=80):
    print('Go swimming in Lake Texoma')
elif(temp>=60):
    print('play tennis')
elif(temp>=40):
    print('play golf')
else:
    print('go skiing on the Grand Ave Overpass')

What is the temperature?  15


visit our campus bookstore!


8. (10 points) Using the `random` package, simulate flipping a fair coin at random and repeat it 1000 times. How many times were heads flipped?

In [37]:
## Answer

heads = 0
tails = 0

import random

for i in range(1000):
    flip = random.randrange(0,2)
    
    if(flip==0):
        heads += 1
    else:
        tails += 1

print(heads)

502


> Answer: 502

### Advanced (20 points)

9. (10 points) Cosine can be estimated as the sum of n terms as follows:
$$
\cos(x) = \sum_{i=0}^{n} \frac{(-1)^i * x^{2i} }{(2i)!}
$$

    In other words, we can approximate the value of $\cos(x)$ by the following sum of terms:
$$
\cos(x) = 1 - \frac{x^2}{2!} + \frac{x^4}{4!} - \frac{x^6}{6!} + ... 
$$

We can calculate this with Python with a "summing-as-you-go-strategy" one term at a time ... as shown by the following pseudo-code :
* initialize summing variable to zero.
* initialize i to 0
* repeat n times 
   - calculate the current term given the current value of i   
   - add the current term to summing variable
   - add 1 to i
   
Provide a Python implementation of the cos series sum above.  Prompt the user for x and n (the number of terms you would like to use for estimation).  Then compute the estimate of cosine using n=10 terms.Print out the results.   **Also** print what the math library calculates for cos(x).  Calculate and print the difference.

In [44]:
## Answer
import math

x = int(input('Value: '))
n = int(input('Iterations: '))

sum = 0

for i in range(n):
    sum += ((-1)**i * (x**(2*i)))/math.factorial(i*2)

print(sum)

Value:  1
Iterations:  10


0.5403023058681397


10. (10 points) The area of a square whose width is 2r and height is 2r is easy to calculate.

    $$
    A_{square} = width * height = 2r * 2r = 4*r^2
    $$

    The formula for the Area of a circle with radius $r$ is given by the following:

    $$
A_{circle} = \pi * r^2 
    $$

    The ratio of the areas is given by:

    $${A_{circle} \over A_{square}} = {(\pi * r^2)\over(4*r^2)} = {\pi \over 4}$$

    Now, it turns out we can approximate the ratio $A_{circle} \over A_{square}$ by randomly sampling (picking random points in the square) and testing if the point is INSIDE the circle.  If we do this many times and we can count the number of times our random point is INSIDE the circle we can estimate the ratio using the ratio $ count_{inside} \over count_{total}$. The count ratio will approach the actual ratio as the number of samples gets large (as the number of iterations approaches infinity).

    <c>![Image](estimating-pi-monte-carlo-method.png)</c>

    Therefore, if we repeatedly sample our space with random points and count the number of points INSIDE the circle.   we can calculate the ratio.

    $${count_{circle} \over count_{total}} \approx {\pi \over 4}$$

    or 

    $$\pi \approx 4 * {count_{circle} \over count_{total}}$$

    Write a python algorithm that prompts the user for the value of n and derives/calculates an approximation of $\pi$ after n iterations.   The algorithm should repeatedly use the random library 2 times to "pick" a random point (x,y) where the x value is a random between -1 and +1.  And the y value is also a random between -1 and +1.  With x,y calculate the distance from the origin.     

    $$
    d = \sqrt{ x^2 + y^2 }
    $$

    Compare your distance to the radius ($r=1$).  if d is less than or equal to one, the sample point is in/on the circle.  

    If you repeat this over and over n times, you can keep count of the number of samples landing *inside* or *on* the circle.  After n times, calculate the final answer using the equation above (4 times the ratio circle count / n ).

    Print your final answer. It is close to $\pi = 3.14159$?   

In [25]:
## Answer
import random
import math

n = int(input('Iterations: '))
cCount = 0
sCount = 0

for i in range(n):
    x = random.uniform(-1, 1)
    y = random.uniform(-1, 1)
    d = math.sqrt(x**2 + y**2)
    
    if(d<=1):
        cCount += 1
    else:
        sCount += 1

print((4*cCount)/(sCount+cCount))

Iterations:  100000


3.137
